contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,212 @@
+
+contract TokenDistributor is ITokenDistributor {
+    /// @notice represents the status of if a person already
+    /// claimed their epoch
+    mapping(address => mapping(uint => bool)) internal claimedEpochs;
+
+    /// @notice kwenta interface
+    IKwenta private kwenta;
+
+    /// @notice rewards staking contract
+    StakingRewardsV2 private stakingRewardsV2;
+
+    /// @notice escrow contract which holds (and may stake) reward tokens
+    RewardEscrowV2 private rewardEscrowV2;
+
+    /// @notice last recorded balance of KWENTA in contract
+    uint internal lastTokenBalance;
+
+    /// @notice last checkpoint time
+    uint internal lastCheckpoint;
+
+    /// @notice starting week of deployment
+    uint internal startTime;
+
+    /// @inheritdoc ITokenDistributor
+    uint[1000000000000000] public tokensPerEpoch;
+
+    /// @notice the week offset in seconds
+    uint internal offset;
+
+    /// @notice max amount of days the epoch can be offset by
+    uint internal constant MAX_OFFSET_DAYS = 6;
+
+    /// @notice weeks in a year
+    uint internal constant WEEKS_IN_YEAR = 52;
+
+    /// @notice constructs the TokenDistributor contract
+    /// and sets startTime
+    /// @param _kwenta: address of the kwenta contract
+    /// @param _stakingRewardsV2: address of the stakingRewardsV2 contract
+    /// @param _rewardEscrowV2: address of the rewardEscrowV2 contract
+    constructor(
+        address _kwenta,
+        address _stakingRewardsV2,
+        address _rewardEscrowV2,
+        uint daysToOffsetBy
+    ) {
+        kwenta = IKwenta(_kwenta);
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+
+        /// @notice custom start day (startTime + daysToOffsetBy)
+        if (daysToOffsetBy > MAX_OFFSET_DAYS) {
+            revert OffsetTooBig();
+        }
+        offset = daysToOffsetBy * 1 days;
+        uint startOfThisWeek = _startOfWeek(block.timestamp);
+        startTime = startOfThisWeek;
+        lastCheckpoint = startOfThisWeek;
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function checkpointToken() public override {
+        uint tokenBalance = kwenta.balanceOf(address(this));
+        uint toDistribute = tokenBalance - lastTokenBalance;
+        lastTokenBalance = tokenBalance;
+
+        uint previousCheckpoint = lastCheckpoint;
+        uint sinceLast = block.timestamp - previousCheckpoint;
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52 weeks
+        for (uint i = 0; i < WEEKS_IN_YEAR; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
```

Comment:
**Thought:** Is it not correct that if `sinceLast == 0` then `block.timestamp == previousCheckpoint` by definition. Therefore only one of these conditions is needed?

Reply:
Actually I think I am wrong here, mybad.